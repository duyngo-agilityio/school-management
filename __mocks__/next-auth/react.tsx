import { createContext } from "react";

export const useSession = () => ({
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
});

export type Session = {
  data: {
    user: {
      token: string;
      email: string;
      id: string;
      name: string;
    };
    expires: string;
  };
  status: "loading" | "authenticated" | "unauthenticated";
  update: (data?: Session) => Promise<Session | null>;
};

const INITIAL_SESSION: Session = {
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
};

const SessionContext = createContext(INITIAL_SESSION);

export const SessionProvider = ({ children }: React.PropsWithChildren) => {
  <SessionContext.Provider value={INITIAL_SESSION}>
    {children}
  </SessionContext.Provider>;
};
