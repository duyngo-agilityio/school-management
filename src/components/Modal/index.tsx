import {
  Modal as BaseModal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  ModalProps,
} from '@chakra-ui/react';
import React, { ReactNode } from 'react';

interface BaseModalProps extends ModalProps {
  isOpen: boolean;
  title: string;
  children: ReactNode;
  onClose: () => void;
}

const Modal = ({
  isOpen,
  title,
  children,
  onClose,
  ...rest
}: BaseModalProps) => {
  return (
    <BaseModal
      closeOnOverlayClick={false}
      isOpen={isOpen}
      onClose={onClose}
      {...rest}
    >
      <ModalOverlay />
      <ModalContent minWidth="fit-content" height="fit-content">
        <ModalHeader>{title}</ModalHeader>
        <ModalCloseButton />
        <ModalBody pb={6}>{children}</ModalBody>
      </ModalContent>
    </BaseModal>
  );
};

export default Modal;
