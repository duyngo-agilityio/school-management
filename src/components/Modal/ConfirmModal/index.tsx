// Components
import { Button, Flex, Text } from '@chakra-ui/react';
import Modal from '..';

interface ConfirmModalProps {
  title: string;
  subTitle?: string;
  isOpen: boolean;
  onClose: () => void;
  onSubmit: () => void;
}

const ConfirmModal = ({
  title,
  subTitle,
  isOpen,
  onClose,
  onSubmit,
}: ConfirmModalProps) => {
  return (
    <Modal isOpen={isOpen} onClose={onClose} title={title} size="xs">
      <Text variant="primary" size="md">
        {subTitle}
      </Text>
      <Flex justifyContent="flex-end" mt="30px">
        <Button mr="10px" onClick={onClose}>
          Cancel
        </Button>
        <Button colorScheme="red" onClick={onSubmit}>
          Delete
        </Button>
      </Flex>
    </Modal>
  );
};

export default ConfirmModal;
