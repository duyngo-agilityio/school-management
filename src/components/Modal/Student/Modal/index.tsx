'use client';

// Libs
import { Controller, useForm } from 'react-hook-form';
import { useRouter } from 'next/navigation';

// Types
import { IStudent } from '@/types';

// Constants
import {
  ERROR_MESSAGES,
  OPTIONS_CLASS,
  OPTIONS_GENDER,
  REGEX_EMAIL,
  REGEX_PASSWORD,
  REGEX_PHONE_NUMBER,
  ROUTES,
  SUCCESS_MESSAGES,
  VALIDATE_MESSAGE,
} from '@/constants';

// Components
import Modal from '../..';
import { Dropdown, PasswordInput, UploadImage } from '@/components';
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
  NumberDecrementStepper,
  NumberIncrementStepper,
  NumberInput,
  NumberInputField,
  NumberInputStepper,
  Textarea,
  useToast,
} from '@chakra-ui/react';

// Services
import { addStudent, editStudent } from '@/actions';

interface StudentModalProps {
  onClose: () => void;
  defaultValues?: IStudent;
}

const StudentModal = ({ defaultValues, onClose }: StudentModalProps) => {
  const { push } = useRouter();
  const toast = useToast();
  const {
    handleSubmit,
    control,
    setValue,
    watch,
    formState: { isSubmitting, isValid, isDirty },
  } = useForm<IStudent>({
    defaultValues,
    mode: 'onBlur',
  });

  const watchAvatar = watch('avatar');

  const onSubmit = async (data: IStudent) => {
    const {
      id,
      fullName,
      className,
      gender,
      age,
      email,
      phoneNumber,
      password,
      description,
    } = data;

    const payload = {
      ...data,
      fullName,
      className,
      gender: Number(gender),
      age: Number(age),
      email,
      phoneNumber,
      password,
      description,
    };

    const dataResponse = defaultValues
      ? await editStudent(payload, id)
      : await addStudent(payload);

    onClose();

    push(ROUTES.STUDENT);

    toast({
      title: dataResponse
        ? defaultValues
          ? SUCCESS_MESSAGES.EDIT_STUDENT
          : SUCCESS_MESSAGES.ADD_STUDENT
        : defaultValues
          ? ERROR_MESSAGES.EDIT_STUDENT
          : ERROR_MESSAGES.ADD_STUDENT,
      status: dataResponse ? 'success' : 'error',
    });
  };

  return (
    <Modal
      onClose={onClose}
      title={defaultValues ? 'Edit Student' : 'Add Student'}
      size="4xl"
    >
      <form onSubmit={handleSubmit(onSubmit)}>
        <Flex flexDirection="row" justifyContent="space-between">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="fullName"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Full name</FormLabel>
                  <Input {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Flex>
            <Controller
              control={control}
              name="className"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12} mr="45px">
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    options={OPTIONS_CLASS}
                    placeholder="Class"
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
            <Controller
              control={control}
              name="gender"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12}>
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    placeholder="Gender"
                    options={OPTIONS_GENDER}
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Flex>
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="email"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_EMAIL,
                  message: VALIDATE_MESSAGE.EMAIL,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Email address</FormLabel>
                  <Input {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Controller
            control={control}
            name="phoneNumber"
            rules={{
              required: VALIDATE_MESSAGE.EMPTY,
              pattern: {
                value: REGEX_PHONE_NUMBER,
                message: VALIDATE_MESSAGE.PHONE_NUMBER,
              },
            }}
            render={({ field, fieldState: { error } }) => (
              <FormControl mt={4}>
                <FormLabel>Phone number</FormLabel>
                <Input {...field} />
                {error?.message && (
                  <FormHelperText color="red.400">
                    {error.message}
                  </FormHelperText>
                )}
              </FormControl>
            )}
          />
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="400px">
            <Controller
              control={control}
              name="password"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_PASSWORD,
                  message: VALIDATE_MESSAGE.PASSWORD,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Password</FormLabel>
                  <PasswordInput {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400" width="inherit">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
            <Controller
              control={control}
              name="age"
              render={({
                field: { ref, ...restField },
                fieldState: { error },
              }) => (
                <FormControl mt={4}>
                  <FormLabel>Age</FormLabel>
                  <NumberInput {...restField} defaultValue={16}>
                    <NumberInputField ref={ref} name={restField.name} />
                    <NumberInputStepper>
                      <NumberIncrementStepper />
                      <NumberDecrementStepper />
                    </NumberInputStepper>
                  </NumberInput>
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
            <Controller
              control={control}
              name="description"
              render={({ field }) => (
                <FormControl mt={4}>
                  <FormLabel>About</FormLabel>
                  <Textarea {...field} size="sm" />
                </FormControl>
              )}
            />
          </Box>
          <Box w="400px">
            <Controller
              control={control}
              name="avatar"
              render={({ field: { value } }) => {
                const handleChangeImage = (url: string) => {
                  setValue('avatar', url);
                };

                const handleRemoveImage = () => {
                  setValue('avatar', '');
                };

                return (
                  <FormControl mt={4}>
                    <FormLabel htmlFor="avatar">Upload Image</FormLabel>
                    <UploadImage
                      src={value}
                      title="Upload Image"
                      width="full"
                      height="255px"
                      boxHeightImage="255px"
                      alt="Avatar Image"
                      onChange={handleChangeImage}
                      onRemove={handleRemoveImage}
                    />
                  </FormControl>
                );
              }}
            />
          </Box>
        </Flex>

        <Flex mt={5} justifyContent="flex-end">
          <Button onClick={onClose} mr={3}>
            Cancel
          </Button>
          <Button
            isLoading={isSubmitting}
            type="submit"
            colorScheme="blue"
            isDisabled={!isDirty || !isValid || !watchAvatar}
          >
            {defaultValues ? 'Edit Student' : 'Add Student'}
          </Button>
        </Flex>
      </form>
    </Modal>
  );
};

export default StudentModal;
