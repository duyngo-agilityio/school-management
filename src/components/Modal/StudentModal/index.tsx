'use client';

// Libs
import { Controller, useForm } from 'react-hook-form';

// Types
import { IStudent } from '@/types';

// Constants
import { OPTIONS_GENDER, OPTIONS_SUBJECT, VALIDATE_MESSAGE } from '@/constants';

// Components
import Modal from '..';
import { Dropdown, PasswordInput } from '@/components';
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
} from '@chakra-ui/react';

interface StudentModalProps {
  title: string;
  onClose: () => void;
  isOpen: boolean;
}

const StudentModal = ({ title, onClose, isOpen }: StudentModalProps) => {
  const { handleSubmit, control } = useForm<IStudent>({
    defaultValues: {
      fullName: '',
      email: '',
      gender: 0,
      password: '',
      phoneNumber: '',
    },
    mode: 'onBlur',
  });

  // TODO: Implement later
  const onSubmit = () => {};

  return (
    <Modal isOpen={isOpen} onClose={onClose} title={title} size="4xl">
      <form onSubmit={handleSubmit(onSubmit)}>
        <Flex flexDirection="row" justifyContent="space-between">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="fullName"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={4}>
                  <FormLabel>Full name</FormLabel>
                  <Input value={value} onChange={onChange} />
                  <FormErrorMessage>{error && error.message}</FormErrorMessage>
                </FormControl>
              )}
            />
          </Box>
          <Flex>
            <Controller
              control={control}
              name="gender"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({ field: { onChange }, fieldState: { error } }) => (
                <Box mt={12} mr="45px">
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    options={OPTIONS_SUBJECT}
                  />
                  <FormErrorMessage>{error && error.message}</FormErrorMessage>
                </Box>
              )}
            />
            <Controller
              control={control}
              name="gender"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({ field: { onChange }, fieldState: { error } }) => (
                <Box mt={12}>
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    placeholder="Gender"
                    options={OPTIONS_GENDER}
                  />
                  <FormErrorMessage>{error && error.message}</FormErrorMessage>
                </Box>
              )}
            />
          </Flex>
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="email"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={4}>
                  <FormLabel>Email address</FormLabel>
                  <Input value={value} onChange={onChange} />
                  <FormErrorMessage>{error && error.message}</FormErrorMessage>
                </FormControl>
              )}
            />
          </Box>
          <Controller
            control={control}
            name="phoneNumber"
            render={({ field: { value, onChange }, fieldState: { error } }) => (
              <FormControl mt={4}>
                <FormLabel>Phone number</FormLabel>
                <Input value={value} onChange={onChange} />
                <FormErrorMessage>{error && error.message}</FormErrorMessage>
              </FormControl>
            )}
          />
        </Flex>
        <Controller
          control={control}
          name="password"
          rules={{
            required: VALIDATE_MESSAGE.EMPTY,
          }}
          render={({ fieldState: { error } }) => (
            <FormControl mt={4}>
              <FormLabel>Password</FormLabel>
              <PasswordInput />
              <FormErrorMessage>{error && error.message}</FormErrorMessage>
            </FormControl>
          )}
        />
        <Flex mt={5} justifyContent="flex-end">
          <Button onClick={onClose} mr={3}>
            Cancel
          </Button>
          <Button type="submit" colorScheme="blue">
            Add Student
          </Button>
        </Flex>
      </form>
    </Modal>
  );
};

export default StudentModal;
