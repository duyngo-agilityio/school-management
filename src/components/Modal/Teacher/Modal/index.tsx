'use client';

// Libs
import { Controller, useForm } from 'react-hook-form';

// Types
import { ITeacher } from '@/types';

// Constants
import {
  ERROR_MESSAGES,
  OPTIONS_CLASS,
  OPTIONS_GENDER,
  OPTIONS_SUBJECT,
  REGEX_EMAIL,
  REGEX_PASSWORD,
  REGEX_PHONE_NUMBER,
  SUCCESS_MESSAGES,
  VALIDATE_MESSAGE,
} from '@/constants';

// Components
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
  NumberDecrementStepper,
  NumberIncrementStepper,
  NumberInput,
  NumberInputField,
  NumberInputStepper,
  Textarea,
  useToast,
} from '@chakra-ui/react';
import { Dropdown, PasswordInput, UploadImage } from '@/components';
import Modal from '../..';

// Services
import { addTeacher, editTeacher } from '@/actions';

interface TeacherModalProps {
  onClose: () => void;
  defaultValues?: ITeacher;
}

const TeacherModal = ({ onClose, defaultValues }: TeacherModalProps) => {
  const toast = useToast();
  const {
    handleSubmit,
    control,
    setValue,
    watch,
    formState: { isDirty, isValid, isSubmitting },
  } = useForm<ITeacher>({
    defaultValues,
    mode: 'onBlur',
  });

  const watchAvatar = watch('avatar');

  const onSubmit = async (data: ITeacher) => {
    const {
      id,
      email,
      fullName,
      age,
      gender,
      subject,
      className,
      avatar,
      description,
      password,
      phoneNumber,
    } = data;

    const payload = {
      ...data,
      fullName,
      email,
      age,
      gender,
      subject,
      className,
      avatar,
      description,
      password,
      phoneNumber,
    };

    const dataResponse = defaultValues
      ? await editTeacher(payload, id)
      : await addTeacher(payload);

    onClose();

    toast({
      title: dataResponse
        ? defaultValues
          ? SUCCESS_MESSAGES.EDIT_TEACHER
          : SUCCESS_MESSAGES.ADD_TEACHER
        : defaultValues
          ? ERROR_MESSAGES.EDIT_TEACHER
          : ERROR_MESSAGES.ADD_TEACHER,
      status: dataResponse ? 'success' : 'error',
    });
  };

  return (
    <Modal
      onClose={onClose}
      title={defaultValues ? 'Edit Teacher' : 'Add Teacher'}
      size="4xl"
    >
      <form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="fullName"
          rules={{
            required: VALIDATE_MESSAGE.EMPTY,
          }}
          render={({ field, fieldState: { error } }) => (
            <FormControl mt={4}>
              <FormLabel>Full name</FormLabel>
              <Input {...field} />
              {error?.message && (
                <FormHelperText color="red.400">{error.message}</FormHelperText>
              )}
            </FormControl>
          )}
        />
        <Flex flexDirection="row" justifyContent="space-between">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="email"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_EMAIL,
                  message: VALIDATE_MESSAGE.EMAIL,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Email address</FormLabel>
                  <Input {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Flex>
            <Controller
              control={control}
              name="className"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12} mr="45px">
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    options={OPTIONS_CLASS}
                    placeholder="Class"
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
            <Controller
              control={control}
              name="gender"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12}>
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    placeholder="Gender"
                    options={OPTIONS_GENDER}
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Flex>
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="400px">
            <Controller
              control={control}
              name="password"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_PASSWORD,
                  message: VALIDATE_MESSAGE.PASSWORD,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Password</FormLabel>
                  <PasswordInput {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Box w="400px">
            <Controller
              control={control}
              name="phoneNumber"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_PHONE_NUMBER,
                  message: VALIDATE_MESSAGE.PHONE_NUMBER,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Phone number</FormLabel>
                  <Input {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="full">
            <Box w="full">
              <Controller
                control={control}
                name="age"
                render={({
                  field: { ref, ...restField },
                  fieldState: { error },
                }) => (
                  <FormControl mt={4}>
                    <FormLabel>Age</FormLabel>
                    <NumberInput {...restField} defaultValue={22} min={22}>
                      <NumberInputField ref={ref} name={restField.name} />
                      <NumberInputStepper>
                        <NumberIncrementStepper />
                        <NumberDecrementStepper />
                      </NumberInputStepper>
                    </NumberInput>
                    {error?.message && (
                      <FormHelperText color="red.400">
                        {error.message}
                      </FormHelperText>
                    )}
                  </FormControl>
                )}
              />
            </Box>
            <Box w="full">
              <Controller
                control={control}
                name="subject"
                rules={{
                  required: VALIDATE_MESSAGE.EMPTY,
                }}
                render={({
                  field: { value, onChange },
                  fieldState: { error },
                }) => (
                  <FormControl mt={12}>
                    <Dropdown
                      onChangeValue={onChange}
                      value={value}
                      options={OPTIONS_SUBJECT}
                      placeholder="Subject"
                    />
                    {error?.message && (
                      <FormHelperText color="red.400">
                        {error.message}
                      </FormHelperText>
                    )}
                  </FormControl>
                )}
              />
            </Box>
            <Box w="full">
              <Controller
                control={control}
                name="description"
                render={({ field }) => (
                  <Box mt={7}>
                    <FormControl mt={4}>
                      <FormLabel>About</FormLabel>
                      <Textarea {...field} size="sm" />
                    </FormControl>
                  </Box>
                )}
              />
            </Box>
          </Box>
          <Box w="full">
            <Controller
              control={control}
              name="avatar"
              render={({ field: { value } }) => {
                const handleChangeImage = (url: string) => {
                  setValue('avatar', url);
                };

                const handleRemoveImage = () => {
                  setValue('avatar', '');
                };

                return (
                  <FormControl mt={4}>
                    <FormLabel>Upload Image</FormLabel>
                    <UploadImage
                      title="Upload Image"
                      width="full"
                      height="268px"
                      boxHeightImage="268px"
                      src={value}
                      alt="Avatar Image"
                      onChange={handleChangeImage}
                      onRemove={handleRemoveImage}
                    />
                  </FormControl>
                );
              }}
            />
          </Box>
        </Flex>
        <Flex mt={5} justifyContent="flex-end">
          <Button onClick={onClose} mr={3}>
            Cancel
          </Button>
          <Button
            isLoading={isSubmitting}
            type="submit"
            colorScheme="blue"
            isDisabled={!isDirty || !isValid || !watchAvatar}
          >
            {defaultValues ? 'Edit Teacher' : 'Add Teacher'}
          </Button>
        </Flex>
      </form>
    </Modal>
  );
};

export default TeacherModal;
