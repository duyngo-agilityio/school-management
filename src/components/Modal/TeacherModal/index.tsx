'use client';

// Libs
import { Controller, useForm } from 'react-hook-form';

// Types
import { ITeacher } from '@/types';

// Constants
import {
  OPTIONS_GENDER,
  OPTIONS_SUBJECT,
  REGEX_EMAIL,
  REGEX_PASSWORD,
  VALIDATE_MESSAGE,
} from '@/constants';

// Components
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
} from '@chakra-ui/react';
import { Dropdown, PasswordInput } from '@/components';
import Modal from '..';

interface TeacherModalProps {
  title: string;
  onClose: () => void;
  isOpen: boolean;
}

const TeacherModal = ({ isOpen, onClose, title }: TeacherModalProps) => {
  const { handleSubmit, control } = useForm<ITeacher>({
    defaultValues: {
      fullName: '',
      email: '',
      gender: 0,
      password: '',
      phoneNumber: '',
      subject: '',
    },
    mode: 'onBlur',
  });

  // TODO: Implement later
  const onSubmit = () => {};

  return (
    <Modal isOpen={isOpen} onClose={onClose} title={title} size="4xl">
      <form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="fullName"
          rules={{
            required: VALIDATE_MESSAGE.EMPTY,
          }}
          render={({ field, fieldState: { error } }) => (
            <FormControl mt={4}>
              <FormLabel>Full name</FormLabel>
              <Input {...field} />
              {error?.message && (
                <FormHelperText color="red.400">{error.message}</FormHelperText>
              )}
            </FormControl>
          )}
        />
        <Flex flexDirection="row" justifyContent="space-between">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="email"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_EMAIL,
                  message: VALIDATE_MESSAGE.EMAIL,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Email address</FormLabel>
                  <Input {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Flex>
            <Controller
              control={control}
              name="subject"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12} mr="45px">
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    options={OPTIONS_SUBJECT}
                    placeholder="Class"
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
            <Controller
              control={control}
              name="gender"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
              }}
              render={({
                field: { value, onChange },
                fieldState: { error },
              }) => (
                <FormControl mt={12}>
                  <Dropdown
                    width="179px"
                    onChangeValue={onChange}
                    value={value}
                    placeholder="Gender"
                    options={OPTIONS_GENDER}
                  />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Flex>
        </Flex>
        <Flex flexDirection="row">
          <Box mr="45px" w="full">
            <Controller
              control={control}
              name="password"
              rules={{
                required: VALIDATE_MESSAGE.EMPTY,
                pattern: {
                  value: REGEX_PASSWORD,
                  message: VALIDATE_MESSAGE.PASSWORD,
                },
              }}
              render={({ field, fieldState: { error } }) => (
                <FormControl mt={4}>
                  <FormLabel>Password</FormLabel>
                  <PasswordInput {...field} />
                  {error?.message && (
                    <FormHelperText color="red.400">
                      {error.message}
                    </FormHelperText>
                  )}
                </FormControl>
              )}
            />
          </Box>
          <Controller
            control={control}
            name="phoneNumber"
            render={({ field }) => (
              <FormControl mt={4}>
                <FormLabel>Phone number</FormLabel>
                <Input {...field} />
              </FormControl>
            )}
          />
        </Flex>
        <Controller
          control={control}
          name="subject"
          rules={{
            required: VALIDATE_MESSAGE.EMPTY,
          }}
          render={({ field: { onChange }, fieldState: { error } }) => (
            <Box mt={7}>
              <FormControl mt={4}>
                <Dropdown
                  onChangeValue={onChange}
                  options={OPTIONS_SUBJECT}
                  placeholder="Subject"
                />
                {error?.message && (
                  <FormHelperText color="red.400">
                    {error.message}
                  </FormHelperText>
                )}
              </FormControl>
            </Box>
          )}
        />
        <Flex mt={5} justifyContent="flex-end">
          <Button onClick={onClose} mr={3}>
            Cancel
          </Button>
          <Button type="submit" colorScheme="blue">
            Add Teacher
          </Button>
        </Flex>
      </form>
    </Modal>
  );
};

export default TeacherModal;
