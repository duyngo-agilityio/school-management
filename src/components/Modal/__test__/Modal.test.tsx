import { fireEvent, render, screen } from '@testing-library/react';

// Components
import Modal from '..';

const mockOnClose = jest.fn();

const ModalProps = {
  isOpen: false,
  title: '',
  children: <>Form</>,
  onClose: mockOnClose,
};

const ModalComponent = (overrideProps = {}) =>
  render(<Modal {...ModalProps} {...overrideProps} />);

describe('Modal Component', () => {
  it('should match snapshot for Modal', () => {
    const { container } = ModalComponent();

    expect(container).toMatchSnapshot();
  });

  it('should render the modal when isOpen is true', () => {
    render(
      <Modal title="Test Title" onClose={() => {}}>
        <div>Test Content</div>
      </Modal>,
    );

    expect(screen.getByText('Test Title')).toBeInTheDocument();
    expect(screen.getByText('Test Content')).toBeInTheDocument();
  });

  it('should call onClose when close button is clicked', () => {
    const handleClose = jest.fn();

    render(
      <Modal title="Test Title" onClose={handleClose}>
        <div>Test Content</div>
      </Modal>,
    );

    fireEvent.click(screen.getByRole('button', { name: /close/i }));
    expect(handleClose).toHaveBeenCalledTimes(1);
  });
});
