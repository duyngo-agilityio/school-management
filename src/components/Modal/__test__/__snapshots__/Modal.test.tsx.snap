// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should match snapshot for Modal 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
